### DISCUSSION TITLE
Why can't vscode plug-in be used in browser-only mode?

### DISCUSSION THREAD

**Discussion by dddcxy** [eclipse-theia/theia#13547] [Q&A]
Why can't vscode plug-in be used in browser-only mode?

**Comment by msujew** [comment]
Hey @dddcxy,

Short answer: We haven't implemented it yet due to technical challenges and not enough contributors being interested in supporting it.

Long answer: Supporting web extension is pretty complicated, since it's a completely new way of fetching extension manifest, code and resources. In normal browser mode, the backend downloads and then serves all extension content. Obviously, this can't work for the browser only mode due to a lack of server. This would probably require some sort of dedicated server offering that we just haven't started working on yet.

Maybe @sdirix can add more to this.

**Reply by sdirix** [reply]
Hi!

Technically most of the functionality required to serve VS Code web extensions are already in place in "browser-only" mode, i.e. Theia as of now can already run VS Code web extensions purely within the browser. As @msujew indicated, what's missing is the infrastructure to provide the list and content of VS Code web extensions to serve.

We have a POC were we hard coded a Theia plugin and a VS Code Web extension (Typescript LSP) to be served in Theia browser-only. The POC can be found [here](https://plugins--browseronly-theia.netlify.app/). So as of now it's already possible to serve VS Code web extensions, however it's quite a lot of manual work.

For the future it would be nice to support VS Code Web extensions without requiring additional work from developers.

Edit: More information to the current state of "browser-only" can be found in #12852

**Reply by kachurun** [reply]
Support of vscode extensions added in this PR:

https://github.com/eclipse-theia/theia/pull/17122

But extensions like TS LSP still not works. I'm working on it but would be nice to see the code under [this POC](https://plugins--browseronly-theia.netlify.app/).

@sdirix maybe you know how to find sources for this POC?

**Reply by sdirix** [reply]
The code is currently not available. However it's nothing special, just hard coding the plugin metadata instead of collecting it. The gist of it should already be contained in this PR https://github.com/eclipse-theia/theia/pull/14776

**Reply by kachurun** [reply]
Ok, thank you, but forget about it. I just got it working, and the TS LSP works now!
I just need to polish some stuff and update the PR!

OMG, how I missed it https://github.com/eclipse-theia/theia/pull/14776
